sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,473 @@
+    private final Duration maxReadTime;
+
+    public UnboundedMqttSource(String serverUri, String clientId, String topic,
+                               long maxNumRecords, Duration maxReadTime) {
+      this.serverUri = serverUri;
+      this.clientId = clientId;
+      this.topic = topic;
+      this.maxNumRecords = maxNumRecords;
+      this.maxReadTime = maxReadTime;
+    }
+
+    @Override
+    public UnboundedReader createReader(PipelineOptions options,
+                                        MqttCheckpointMark checkpointMark) {
+      return new UnboundedMqttReader(this, checkpointMark);
+    }
+
+    @Override
+    public List<UnboundedMqttSource> generateInitialSplits(int desiredNumSplits,
+                                                           PipelineOptions options) {
+      List<UnboundedMqttSource> sources = new ArrayList<>();
+      for (int i = 0; i < desiredNumSplits; i++) {
+        // NB: it's important that user understand the impact of MQTT message QoS
+        sources.add(new UnboundedMqttSource(serverUri, clientId + "-" + i, topic, maxNumRecords,
+            maxReadTime));
+      }
+      return sources;
+    }
+
+    @Override
+    public void validate() {
+      Preconditions.checkNotNull(serverUri, "serverUri");
+      Preconditions.checkNotNull(clientId, "clientId");
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return SerializableCoder.of(MqttCheckpointMark.class);
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(byte[].class);
+    }
+
+  }
+
+  private static class UnboundedMqttReader extends UnboundedSource.UnboundedReader
+      implements MqttCallback {
+
+    private UnboundedMqttSource source;
+
+    private MqttCheckpointMark checkpointMark;
+
+    private MqttClient client;
+    private byte[] current;
+    private Instant currentTimestamp;
+
+    private UnboundedMqttReader(UnboundedMqttSource source,
+                                MqttCheckpointMark checkpointMark) {
+      this.source = source;
+      if (checkpointMark != null) {
+        this.checkpointMark = checkpointMark;
+      } else {
+        this.checkpointMark = new MqttCheckpointMark();
+      }
+      this.current = null;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      try {
+        client = new MqttClient(source.serverUri, source.clientId);
+        client.connect();
+        if (source.topic != null) {
+          client.subscribe(source.topic);
+        }
+        client.setCallback(this);
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (current == null) {
+        return false;
+      } else {
+        return true;
+      }
+    }
+
+    @Override
+    public void close() throws IOException {
+      try {
+        if (client != null) {
+          client.disconnect();
```

Comment:
Can disconnect() throw an exception? Is it an issue that in that case you don't close the client?